atc/engine/check_delegate.go

```diff
@@ -162,6 +152,18 @@ func (d *checkDelegate) WaitToRun(ctx context.Context, scope db.ResourceConfigSc
 		}
 
 		return nil, false, nil
+	} else {
+		// rate limit periodic resource checks so worker load (plus load on external
+		// services) isn't too spiky
+		if !d.build.IsManuallyTriggered() && d.plan.IsPeriodic() {
```

Comment:
the rate limiting was originally outside the lock acquisition so that we don't sleep() with the lock held, since that'll prevent manually triggered checks from running. do you think moving it here is worth the tradeoff?

i guess folks don't manually trigger checks very often, but the thought is that when they do they probably want it to happen as quickly as possible. but if this approach is causing bigger problems i'm happy to change it.

side note: if we stop reusing check containers we shouldn't need the locking at all[1], but we'd have to do a bit more work to keep a check cache directory between runs so resources aren't running a fresh `git clone` all the time. we'd also want to verify the performance overhead of creating a fresh check container each time. the initial code change would be simple, at least. https://github.com/concourse/concourse/blob/5b7f0b9b5b1b9e287274e56b1e291be05965458b/atc/exec/check_step.go#L353-L355

[1]: might want to make sure the db code for saving versions is safe without an outer lock

Reply:
Yeah, that's a good point. Doing rate limit after acquiring lock is not a good idea.